### PR TITLE
Fix/Clean up Classifiers, Fix RegularizerL2, Add 2-Layer Net Connectivity Tests

### LINKED ISSUE #10: Implement Backprop
- [x] LayerFullyConnected
- [x] LayerReLU
- [x] ClassifierSoftmax
- [x] Connectivity Testing

### PR DESCRIPTION
**Pull request by hershal**
This PR marks the first time we have a somewhat-functioning, ad-hoc, two-layer, network. This process revealed multiple deficiencies in the code:
1. The Softmax classifier was incorrect when given multiple points (images) in a batch.
The Softmax implementation had two bugs and there were holes in the tests which did not reveal the issues. The first bug was that the gradients were not scaled by the number of input points. The second bug was that the gradients at the indices of the correct classifications were not calculated correctly (it needed an `axis=1`).

2. The L2 Regularizer was not properly scaled. The scaling by 1/2 makes the regularizer's gradients 1 so the regularizer has no influence on the farthest gradient.

There are other additions here:
1. I added `set_batch_labels()` on top of the classifiers so that they conform to the `forward()`, `backward()` API.

2. Documentation Cleanups: I indicated Side-Effects when they occurred and updated the Outputs due to (1).

3. Test Cleanups: The Softmax tests now have test fixtures which allow for easier testing and less duplication.

4. Two Layer Network Tests: This was the goal of the PR until it ballooned when I found the bugs, but the PR does include a backprop test of a two-layer network with a forward and backward API.

This resolves #10.

### CODE REVIEW

File: layers/test/test_two_layer_net.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+from layers.fully_connected import LayerFullyConnected
+from regularizers.l2 import RegularizerL2
+
+from lib.gradient_check import eval_numerical_gradient
+from utils.compose import compose
+
+
+class TestTwoLayerNetFixture(unittest.TestCase):
+    def initConstants(self):
+        pass
```
Comment (EricCrosson) [inline layers/test/test_two_layer_net.py]: Might as well raise the new [NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError) here

File: layers/test/test_two_layer_net.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        self.layer0_activations = LayerReLU()
+        self.layer1 = LayerFullyConnected((self.hidden_layer_size, self.num_classifications))
+        self.layer1_activations = LayerReLU()
+        self.classifier = ClassifierSoftmax((self.num_points, self.num_classifications))
+
+        self.regularizer = RegularizerL2()
+        self.regularizer.addLayer(self.layer0)
+        self.regularizer.addLayer(self.layer1)
+
+        self.forward = compose(self.layer0.forward, self.layer0_activations.forward,
```
Comment (EricCrosson) [inline layers/test/test_two_layer_net.py:36]: Any way we can simply write

```python
self.forward = compose(self.layer0, self.layer0_activations, ...)
```

as the `.forward`-s are redundant?

File: classifiers/test/test_softmax.py [hunk 3]
```diff
@@ -14,7 +14,23 @@
 # TODO: clean this up! There's so much duplicated code below...
 
 
-class TestClassifierSoftmax(unittest.TestCase):
+class ClassifierSoftmaxGradientTest():
+    def testGradientNaive(self):
```
Comment (EricCrosson) [inline classifiers/test/test_softmax.py]: This is turning into a sick DSL

File: layers/test/test_two_layer_net.py [hunk 4]
```diff
@@ -68,5 +68,12 @@ def test_forward(self):
...
               15.05099822, 15.66781506, 16.2846319]])
+
+        forward = compose(self.layer0.forward, self.layer0_activations.forward,
+                          self.layer1.forward, self.layer1_activations.forward)
+
         scores = forward(self.forward_input)
         self.assertTrue(np.allclose(scores, correct_scores))
+
+        self.classifier.set_batch_labels(np.array([0, 5, 1]))
+        self.assertAlmostEqual(self.forward(self.forward_input), 3.4702243556)
```
Comment (EricCrosson) [inline layers/test/test_two_layer_net.py]: So this test here is testing our feed-forward ability to... classify?

Comment (hershal) [reply layers/test/test_two_layer_net.py]: Yes, it tests the network's ability to classify and obtain a loss.

File: layers/test/test_two_layer_net.py (hunk 2 above)
Comment (hershal) [reply layers/test/test_two_layer_net.py:36]: Yeah I'll get to that when I write the pipeline API

File: layers/test/test_two_layer_net.py (hunk 1 above)
Comment (hershal) [reply layers/test/test_two_layer_net.py]: Good point.

Comment (hershal) [reply layers/test/test_two_layer_net.py]: I ended up not using `NotImplementedError` because otherwise `initNetwork` wouldn't work by itself, but I'll keep the `NotImplementedError` in mind.

**Review by EricCrosson — APPROVED** [review]
